internal/internal_event_handlers.go

```diff
@@ -384,6 +384,53 @@ func validateAndSerializeSearchAttributes(attributes map[string]interface{}) (*c
 	return attr, nil
 }
 
+func (wc *workflowEnvironmentImpl) UpsertMemo(memo_map map[string]interface{}) error {
+	// This has to be used in WorkflowEnvironment implementations instead of in Workflow for testsuite mock purpose.
+	memo, err := validateAndSerializeMemo(memo_map, wc.dataConverter)
+	if err != nil {
+		return err
+	}
+
+	changeID := wc.GenerateSequenceID()
+	mergedMemo := mergeMemo(wc.workflowInfo.Memo, memo)
+	wc.commandsHelper.modifyProperties(changeID, mergedMemo)
+	// this is for getInfo correctness
+	wc.updateWorkflowInfoWithProperties(memo)
+	return nil
+}
+
+func (wc *workflowEnvironmentImpl) updateWorkflowInfoWithProperties(memo *commonpb.Memo) {
```

Comment:
I'm naming it generic because so it can be extended later when new attributes are added.
I can make it specific, but I suppose new attributes would require creating individual functions to update each of them.

Reply:
The parameter is very specific though, so it's obviously not capable of future attributes without altering the signature anyways. So might as well not have a generic name w/ one specific param for the one specific use case.